cpp/include/cudf/ast/detail/operators.hpp

```diff
@@ -780,6 +789,51 @@ struct operator_functor<ast_operator::NOT, false> {
   }
 };
 
+// Type traits used to determine when a casting operation is valid.
+template <typename T, typename U, typename = void>
+struct can_cast : std::false_type {
+};
+
+template <typename T, typename U>
+struct can_cast<T, U, std::void_t<decltype(static_cast<U>(std::declval<T>()))>> : std::true_type {
+};
+
+template <typename T, typename U>
+constexpr bool can_cast_v = can_cast<T, U>::value;
+
+template <>
+struct operator_functor<ast_operator::CAST_TO_INT64, false> {
+  static constexpr auto arity{1};
+
+  template <typename InputT, std::enable_if_t<can_cast_v<InputT, int64_t>>* = nullptr>
+  CUDA_DEVICE_CALLABLE auto operator()(InputT input) -> int64_t
+  {
+    return static_cast<int64_t>(input);
+  }
+};
+
+template <>
+struct operator_functor<ast_operator::CAST_TO_UINT64, false> {
+  static constexpr auto arity{1};
+
+  template <typename InputT, std::enable_if_t<can_cast_v<InputT, uint64_t>>* = nullptr>
+  CUDA_DEVICE_CALLABLE auto operator()(InputT input) -> uint64_t
+  {
+    return static_cast<uint64_t>(input);
+  }
```

Comment:
trailing return type with decltype will act like SFINAE.
For only convertible types, this function is defined.

```suggestion
  template <typename InputT>
  CUDA_DEVICE_CALLABLE auto operator()(InputT input) ->  decltype(static_cast<uint64_t>(input))
  {
    return static_cast<uint64_t>(input);
  }
```

Reply:
Expression SFINAE https://en.cppreference.com/w/cpp/language/sfinae